common/bitcointree/musig2.go

```diff
@@ -585,57 +586,61 @@ type readable interface {
 func encodeMatrix[T writable](matrix [][]T) ([]byte, error) {
 	var buf bytes.Buffer
 
+	// Write number of rows
+	if err := binary.Write(&buf, binary.LittleEndian, uint32(len(matrix))); err != nil {
+		return nil, err
+	}
+
+	// For each row, write its length and then its elements
 	for _, row := range matrix {
+		// Write row length
+		if err := binary.Write(&buf, binary.LittleEndian, uint32(len(row))); err != nil {
+			return nil, err
+		}
+		// Write row data
 		for _, cell := range row {
-			if err := buf.WriteByte(columnSeparator); err != nil {
-				return nil, err
-			}
-
 			if err := cell.Encode(&buf); err != nil {
 				return nil, err
 			}
 		}
-
-		if err := buf.WriteByte(rowSeparator); err != nil {
-			return nil, err
-		}
 	}
 
 	return buf.Bytes(), nil
 }
 
 // decodeMatrix decode a byte stream into a matrix of serializable objects
 func decodeMatrix[T readable](factory func() T, data io.Reader) ([][]T, error) {
-	matrix := make([][]T, 0)
-	row := make([]T, 0)
-
-	for {
-		separator := make([]byte, 1)
+	var rowCount uint32
 
-		if _, err := data.Read(separator); err != nil {
-			if err == io.EOF {
-				break
-			}
-			return nil, err
-		}
+	// Read number of rows
+	if err := binary.Read(data, binary.LittleEndian, &rowCount); err != nil {
+		return nil, err
+	}
 
-		b := separator[0]
+	// Initialize matrix
+	matrix := make([][]T, rowCount)
 
-		if b == rowSeparator {
-			matrix = append(matrix, row)
-			row = make([]T, 0)
-			continue
+	// For each row, read its length and then its elements
+	for i := uint32(0); i < rowCount; i++ {
+		var colCount uint32
+		// Read row length
+		if err := binary.Read(data, binary.LittleEndian, &colCount); err != nil {
+			return nil, err
 		}
 
-		cell := factory()
+		// Initialize row
+		row := make([]T, colCount)
 
-		if err := cell.Decode(data); err != nil {
-			if err == io.EOF {
-				break
+		// Read row data
+		for j := uint32(0); j < colCount; j++ {
+			cell := factory()
+			if err := cell.Decode(data); err != nil {
+				return nil, err
 			}
-			return nil, err
+			row[j] = cell
 		}
-		row = append(row, cell)
+
+		matrix[i] = row
```

Comment:
Please, use append instead of assingment to populate slice

```suggestion
	matrix := make([][]T, 0, rowCount)

	// For each row, read its length and then its elements
	for i := uint32(0); i < rowCount; i++ {
		var colCount uint32
		// Read row length
		if err := binary.Read(data, binary.LittleEndian, &colCount); err != nil {
			return nil, err
		}

		// Initialize row
		row := make([]T, 0, colCount)

		// Read row data
		for j := uint32(0); j < colCount; j++ {
			cell := factory()
			if err := cell.Decode(data); err != nil {
				return nil, err
			}
			row = append(row, cell)
		}

		matrix = append(matrix, row)
```

Reply:
done